### PR TITLE
refactor: class instead of var+interface in d.ts

### PR DESCRIPTION
**Pull request by lucacasonato**
This dramatically improves docs:

1.4
image

This PR
image

cc @kitsonk

### CODE REVIEW

**Comment by ry** [comment]
I believe this copied this way out of typescript's lib.dom.d.ts... I'm fine with changing it.

**Comment by solson** [comment]
I think this fixes https://github.com/denoland/deno/issues/7472.

**Review by kitsonk — APPROVED** [review]
We already have a challenge with trying to keep it in sync with `lib.dom.d.ts`, so adjusting this I don't see as a huge problem.  We need a way of trying to detect mis-alignment with `lib.dom.d.ts` but that is a different issue.

It just has a conflict with master that needs to be changed.  Also we might want to make sure that any impacts on #7468 are addressed.

File: .dprintrc.json [hunk 1]
```diff
@@ -10,10 +10,22 @@
...
   },
-  "includes": ["**/*.{ts,tsx,js,jsx,json,md}"],
+  "includes": [
+    "**/*.{ts,tsx,js,jsx,json,md}",
+    "cli/dts/lib.deno.ns.d.ts",
+    "cli/dts/lib.deno.shared_globals.d.ts",
+    "cli/dts/lib.deno.unstable.d.ts",
+    "cli/dts/lib.deno.window.d.ts",
+    "cli/dts/lib.deno.worker.d.ts"
+  ],
```
Comment (nayeemrmn) [inline .dprintrc.json:20]: I think these `d.ts` files are included in the glob anyway.

Comment (lucacasonato) [reply .dprintrc.json]: Yup :+1:

**Comment by kitsonk** [comment]
@orta @DanielRosenwasser I know everyone is super busy, but can either of you think of any disadvantages/challenges with declaring the parts of `dom` we support like this?  I can't, but wouldn't mind a more informed opinion.

**Comment by kitsonk** [comment]
Ah, I just remembered a big "disadvantage" to this approach, interfaces are open ended, classes are not. This means enhancing or enriching becomes more challenging. This shouldn't be an issue with Deno person, but if people decide to extend the built-ins in Deno, they might run into challenges.

**Comment by nayeemrmn** [comment]
> interfaces are open ended, classes are not

The thing is that the runtime structures that are being represented here _are_ (effectively) classes, without meaningful associated interfaces. As in a mocked WebSocket that merely satisfies the fields of the "WebSocket interface" won't be able to function as one in any host environment API. It must come from the WebSocket constructor. So the interfaces aren't generally helpful.

**Comment by samal-rasmussen** [comment]
> 
> 
> Ah, I just remembered a big "disadvantage" to this approach, interfaces are open ended, classes are not. This means enhancing or enriching becomes more challenging. This shouldn't be an issue with Deno person, but if people decide to extend the built-ins in Deno, they might run into challenges.

That is a huge drawback. JS objects are always actually mutable so things can always be monkey patched down the line. The new class based types makes this unrepresentable in the type system.

**Pull request merged**